lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -1557,36 +1578,129 @@ void OpenCLFunction::freeDeviceBuffer(cl_mem buf) { clReleaseMemObject(buf); }
 void OpenCLFunction::collectConstants(Module *module) {
   runtimeBundle_.collectConstants(module);
 }
+
+void OpenCLFunction::translateTraceEvents(ExecutionContext *context) const {
+  if (context->getTraceContext() == nullptr ||
+      (!getTraceInfo().enabled && !clDoProfile)) {
+    return;
+  }
+  cl_ulong total = 0;
+
+  std::unordered_map<std::string, cl_ulong> kernelToDuration;
+  auto &traceEvents = context->getTraceContext()->getTraceEvents();
+  int tid = context->getTraceContext()->getTraceThread();
+  std::vector<cl_ulong> manualTimestamps;
+
+  for (auto &kl : kernelLaunches_) {
+    auto &event = kl.event_;
+    if (event == nullptr) {
+      continue;
+    }
+    clWaitForEvents(1, &event);
+
+    auto &name = kl.name_;
+    auto &type = kl.type_;
+    assert(!name.empty() && "Kernel name cannot be empty");
+    cl_ulong time_start;
+    cl_ulong time_end;
+
+    clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_START,
+                            sizeof(time_start), &time_start, NULL);
+    clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_END, sizeof(time_end),
+                            &time_end, NULL);
+
+    if (type == "checkpoint") {
+      const auto &it = manualTraceEvents_.find(name);
+      if (it == manualTraceEvents_.end()) {
+        llvm::outs() << "warning: found manual trace event (" << name
```

Comment:
Do we really want to print it each time to the standard output? Should be DEBUG_GLOW()? 

Or may be we should introduce something like a general-purpose way to produce messages in Glow? E.g. `glowProduceMessage(MessageKind, format, ...)`, where MessageKind is an enum defining INFO, ERROR, WARNING, etc?

Reply:
sounds suspiciously like glog...

I'll change it to DEBUG_GLOW for this PR.